packages/protocol/contracts/layer1/preconf/libs/LibPreconfUtils.sol

```diff
@@ -6,18 +6,27 @@ import "./LibPreconfConstants.sol";
 /// @title LibPreconfUtils
 /// @custom:security-contact security@taiko.xyz
 library LibPreconfUtils {
+    uint256 private constant MAX_NUM_QUERIES = 32;
     /// @notice Retrieves the beacon block root for a given timestamp.
-    /// @dev At block N, this function gets the beacon block root for block N - 1.
-    ///      To obtain the block root of the Nth block, it queries the root at block N + 1.
-    ///      If N + 1 is a missed slot, it continues querying until it finds a block N + x
-    ///      that has the block root for the Nth block.
+    /// @dev To obtain the block root of the Nth block, this function queries the root at block N +
+    /// 1. If block N + 1 is a missed slot, it continues querying up to 32 subsequent blocks (N + 2,
+    /// N + 3, etc.) until it finds a block that contains the root for the Nth block or the target
+    /// timestamp exceeds the current block timestamp.
+    /// @dev Caller should verify the returned value is not 0.
     /// @param timestamp The timestamp for which the beacon block root is to be retrieved.
     /// @return The beacon block root as a bytes32 value.
+
     function getBeaconBlockRoot(uint256 timestamp) internal view returns (bytes32) {
+        if (timestamp < LibPreconfConstants.getGenesisTimestamp(block.chainid)) {
+            return bytes32(0);
+        }
         uint256 targetTimestamp = timestamp + LibPreconfConstants.SECONDS_IN_SLOT;
-        while (true) {
```

Comment:
@AnshuJalan I think I got myself confused by the loop in this function. According to https://eips.ethereum.org/EIPS/eip-4788, the implementation should be as simply as: 

```solidity
    function getBeaconBlockRoot(uint256 timestamp) internal view returns (bytes32) {
        (bool success, bytes memory result) =
            LibPreconfConstants.getBeaconBlockRootContract().staticcall(abi.encode(timestamp));

        if (success && result.length == 32) {
            return abi.decode(result, (bytes32));
        } else {
            return bytes32(0); // or revert
        }
    }

```